pkg/sql/idxrecommendations/idx_recommendations_cache.go

```diff
@@ -0,0 +1,199 @@
+	//return !strings.HasPrefix(strings.ToUpper(fingerprint), "PREPARE ")
+}
+
+// ShouldGenerateIndexRecommendation returns true if there was no generation in the past hour
+// and there is at least 5 executions of the same fingerprint/database/planHash combination.
+func (idxRec *IndexRecCache) ShouldGenerateIndexRecommendation(
+	fingerprint string, planHash uint64, database string,
+) bool {
+	if !sqlstats.SampleIndexRecommendation.Get(&idxRec.st.SV) || !statementCanHaveRecommendation(fingerprint) {
+		return false
+	}
+
+	idxKey := indexRecKey{
+		stmtNoConstants: fingerprint,
+		database:        database,
+		planHash:        planHash,
+	}
+	recInfo, found := idxRec.getIndexRecommendation(idxKey)
+	// If the statement is being executed for the first time,
+	// initialize it.
+	if !found {
+		idxRec.setIndexRecommendations(idxKey, timeutil.Now().Add(-duration.SecsPerDay*1e6), []string{}, 1)
+		recInfo, found = idxRec.getIndexRecommendation(idxKey)
+		// If couldn't initialized, don't generate recommendations.
+		if !found {
+			return false
+		}
+	}
+
+	timeSinceLastGenerated := timeutil.Now().Sub(recInfo.lastGeneratedTs)
+	return recInfo.executionCount >= 5 && timeSinceLastGenerated.Hours() >= 1
+}
+
+// UpdateIndexRecommendations updates the values for index recommendations.
+// If reset is true, a new recommendation was generated, so reset the execution counter and
+// lastGeneratedTs, otherwise just increment the executionCount.
+func (idxRec *IndexRecCache) UpdateIndexRecommendations(
+	fingerprint string, planHash uint64, database string, recommendations []string, reset bool,
+) []string {
+	idxKey := indexRecKey{
+		stmtNoConstants: fingerprint,
+		database:        database,
+		planHash:        planHash,
+	}
+
+	if reset {
+		idxRec.setIndexRecommendations(idxKey, timeutil.Now(), recommendations, 0)
+		return recommendations
+	}
+
+	recInfo, found := idxRec.getIndexRecommendation(idxKey)
+	// If is the first time, we want the lastGeneratedTs to be in the past, in case we reach
+	// the execution count, we should generate new recommendations.
+	generatedTime := timeutil.Now().Add(-duration.SecsPerDay * 1e6)
+	count := int64(1)
+	if found {
+		generatedTime = recInfo.lastGeneratedTs
+		recommendations = recInfo.recommendations
+		count = recInfo.executionCount + 1
+	}
+	idxRec.setIndexRecommendations(idxKey, generatedTime, recommendations, count)
+
+	return recInfo.recommendations
+}
+
+func (idxRec *IndexRecCache) getIndexRecommendation(key indexRecKey) (indexRecInfo, bool) {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
+
+	recInfo, found := idxRec.mu.idxRecommendations[key]
+
+	return recInfo, found
+}
+
+func (idxRec *IndexRecCache) setIndexRecommendations(
+	key indexRecKey, time time.Time, recommendations []string, execCount int64,
+) {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
+
+	_, found := idxRec.getIndexRecommendation(key)
```

Comment:
consider returning early if you found it? A way I've seen this done elsewhere is to have a getOrCreate method which attempts to do the get and then re-locks exclusively and attempts to do the create assuming there was not a race.

Reply:
I can't really return early, this is just looking if the entry exists, then I need to do the actual set of the new value. I did like the idea of the getOrCreate so made some updates to use that (here and in the other functions too)